src/history.rs

```diff
@@ -155,126 +139,231 @@ impl<'a> DynamicHistory<'a> {
-        if self.compact_log.try_push(ev).is_err() {
-            return;
-        }
-        if self.compact_log.try_push(payload).is_ok() {
-            self.event_count = self.event_count.saturating_add(1);
-        }
+        let (first_overwritten, second_overwritten) =
+            self.log.write_event_with_payload(event_id, payload);
+        self.merge_overwritten_clock(first_overwritten);
+        self.merge_overwritten_clock(second_overwritten);
+        self.event_count = self.event_count.saturating_add(1);
     }
 
     /// Increments the clock in the logical clock corresponding to this probe instance
     #[inline]
-    fn increment_local_clock_count(&mut self) {
+    fn increment_local_clock(&mut self) {
         // N.B. We rely on the fact that the first member of the clocks
         // collection is always the clock for this probe
-        debug_assert!(self.probe_id == self.clocks[0].id);
-        debug_assert!(
-            !self.chunked_report_state.is_report_in_progress(),
-            "Should not be incrementing the local clock count when a report is in progress"
-        );
-        self.clocks.as_mut_slice()[0].increment();
+        self.self_clock.increment();
         self.event_count = 0;
     }
 
     #[inline]
-    pub(crate) fn produce_snapshot(&mut self) -> Result<CausalSnapshot, ProduceError> {
-        if self.chunked_report_state.is_report_in_progress() {
-            return Err(ProduceError::ReportLockConflict);
+    pub(crate) fn now(&self) -> ModalityProbeInstant {
+        ModalityProbeInstant {
+            clock: self.self_clock,
+            event_count: self.event_count,
+        }
+    }
+
+    pub(crate) fn report(&mut self, destination: &mut [u8]) -> Result<usize, ReportError> {
+        // If we can't store at least a header and one event, it's a hard error
+        if destination.len() < WireReport::<&[u8]>::buffer_len(0, 1) {
+            return Err(ReportError::InsufficientDestinationSize);
+        }
+
+        let self_clock = self.self_clock;
+        let mut report = WireReport::new_unchecked(destination);
+        report.set_fingerprint();
+        report.set_probe_id(self.probe_id);
+        report.set_clock(crate::pack_clock_word(self_clock.epoch, self_clock.ticks));
+
+        // We can't store at least the frontier clocks and a single two-word item
+        if report.as_ref().len() < WireReport::<&[u8]>::buffer_len(self.clocks.len(), 2) {
+            report.set_seq_num(self.report_seq_num);
+            report.set_n_clocks(0);
+            report.set_n_log_entries(1);
+            let payload = report.payload_mut();
+            payload[..size_of::<LogEntry>()].copy_from_slice(
+                &EventId::EVENT_INSUFFICIENT_REPORT_BUFFER_SIZE
+                    .get_raw()
+                    .to_le_bytes(),
+            );
+        } else {
+            if self.log_items_missed != 0 {
+                self.record_event_with_payload(
+                    EventId::EVENT_LOG_ITEMS_MISSED,
```

Comment:
Should write this into the report log output (after the frontier clocks, before the rest of the log), but not into the probe's self.log.

Reply:
Good catch, fixed in cc1273e